circuits/float/bootstrapper/bootstrapping/bootstrapping.go

```diff
@@ -0,0 +1,364 @@
+	// to the input scale
+	diffScale := ctIn.Scale.Div(ctOut.Scale).Bigint()
+
+	// [M^{d} + e^{d-logprec}]
+	if err = btp.Mul(ctOut, diffScale, ctOut); err != nil {
+		return nil, err
+	}
+	ctOut.Scale = ctIn.Scale
+
+	if btp.IterationsParameters != nil {
+
+		var totLogPrec float64
+
+		for i := 0; i < len(btp.IterationsParameters.BootstrappingPrecision); i++ {
+
+			logPrec := btp.IterationsParameters.BootstrappingPrecision[i]
+
+			totLogPrec += logPrec
+
+			// prec = round(2^{logprec})
+			log2 := bignum.Log(new(big.Float).SetPrec(256).SetUint64(2))
+			log2TimesLogPrec := log2.Mul(log2, new(big.Float).SetFloat64(totLogPrec))
+			prec := new(big.Int)
+			log2TimesLogPrec.Add(bignum.Exp(log2TimesLogPrec), new(big.Float).SetFloat64(0.5)).Int(prec)
+
+			// round(q1/logprec)
+			scale := new(big.Int).Set(diffScale)
+			bignum.DivRound(scale, prec, scale)
+
+			// Checks that round(q1/logprec) >= 2^{logprec}
+			requiresReservedPrime := scale.Cmp(new(big.Int).SetUint64(1)) < 0
+
+			if requiresReservedPrime && btp.IterationsParameters.ReservedPrimeBitSize == 0 {
+				return ctOut, fmt.Errorf("warning: early stopping at iteration k=%d: reason: round(q1/2^{logprec}) < 1 and no reserverd prime was provided", i+1)
+			}
+
+			// [M^{d} + e^{d-logprec}] - [M^{d}] -> [e^{d-logprec}]
+			tmp, err := btp.SubNew(ctOut, ctIn)
+
+			if err != nil {
+				return nil, err
+			}
+
+			// prec * [e^{d-logprec}] -> [e^{d}]
+			if err = btp.Mul(tmp, prec, tmp); err != nil {
+				return nil, err
+			}
+
+			tmp.Scale = ctOut.Scale
+
+			// [e^{d}] / q1 -> [e^{d}/q1]
+			if tmp, errScale, err = btp.scaleDownToQ0OverMessageRatio(tmp); err != nil {
+				return nil, err
+			}
+
+			// [e^{d}/q1] -> [e^{d}/q1 + e'^{d-logprec}]
+			if tmp, err = btp.bootstrap(tmp); err != nil {
+				return nil, err
+			}
+
+			tmp.Scale = tmp.Scale.Mul(*errScale)
+
+			// [[e^{d}/q1 + e'^{d-logprec}] * q1/logprec -> [e^{d-logprec} + e'^{d-2logprec}*q1]
+			// If scale > 2^{logprec}, then we ensure a precision of at least 2^{logprec} even with a rounding of the scale
+			if !requiresReservedPrime {
+				if err = btp.Mul(tmp, scale, tmp); err != nil {
+					return nil, err
+				}
+			} else {
+
+				// Else we compute the floating point ratio
+				ss := new(big.Float).SetInt(diffScale)
+				ss.Quo(ss, new(big.Float).SetInt(prec))
+
+				// Do a scaled multiplication by the last prime
+				if err = btp.Mul(tmp, ss, tmp); err != nil {
+					return nil, err
+				}
+
+				// And rescale
+				if err = btp.Rescale(tmp, tmp); err != nil {
+					return nil, err
+				}
+			}
+
+			// This is a given
+			tmp.Scale = ctOut.Scale
+
+			// [M^{d} + e^{d-logprec}] - [e^{d-logprec} + e'^{d-2logprec}*q1] -> [M^{d} + e'^{d-2logprec}*q1]
+			if err = btp.Sub(ctOut, tmp, ctOut); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return
+}
+
+func currentMessageRatioIsGreaterOrEqualToLastPrimeTimesTargetMessageRatio(ct *rlwe.Ciphertext, msgRatio float64, r *ring.Ring) bool {
```

Comment:
function name is a bit too long in this case, if possible would be better to make it more concise

Reply:
Done